### PR TITLE
FastCV-based HAL for OpenCV acceleration 2ndpost-2

### PR DESCRIPTION
**Pull request by adsha-quic**
### Detailed description:

- Add support for multiply 8u, 16s and 32f
- Add support for cv_hal_pyrdown 8u
- Add support for cv_hal_cvtBGRtoHSV and cv_hal_cvtBGRtoYUVApprox 8u

Requires binary from [opencv/opencv_3rdparty#90](https://github.com/opencv/opencv_3rdparty/pull/90)
Depends on: [opencv/opencv#26617](https://github.com/opencv/opencv/pull/26617)

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [ ] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: 3rdparty/fastcv/src/fastcv_hal_imgproc.cpp [hunk 1]
```diff
@@ -141,28 +221,88 @@ int fastcv_hal_sobel(
...
+                status = fcvFilterSobel5x5u8s16(src_data, src_width, src_height, src_step, dxBuffer, dyBuffer, dst_step, fcvBorder, 0);
+                break;
+            }
+            case FCV_MAKETYPE(7,CV_8U):
+            {
+                status = fcvFilterSobel7x7u8s16(src_data, src_width, src_height, src_step, dxBuffer, dyBuffer, dst_step, fcvBorder, 0);
+                break;
+            }
+            default:
+                break;
```
Comment (asmorkalov) [inline 3rdparty/fastcv/src/fastcv_hal_imgproc.cpp:286]: I propose to add at least CV_Assert here. Just for the case, if invoker code will be changed incorrectly in future.

File: 3rdparty/fastcv/src/fastcv_hal_imgproc.cpp [hunk 2]
```diff
@@ -52,6 +52,97 @@ int fastcv_hal_medianBlur(
...
+            {
+                fcvFilterSobel5x5u8s16(src, width, rangeHeight, src_step, dxBuffer, dyBuffer, dst_step, fcvBorder, 0);
+                break;
+            }
+            case FCV_MAKETYPE(7,CV_8U):
+            {
+                fcvFilterSobel7x7u8s16(src, width, rangeHeight, src_step, dxBuffer, dyBuffer, dst_step, fcvBorder, 0);
+                break;
+            }
+            default:
```
Comment (asmorkalov) [inline 3rdparty/fastcv/src/fastcv_hal_imgproc.cpp:116]: I propose to add at least CV_Assert here. Just for the case, if invoker code will be changed incorrectly in future.

File: 3rdparty/fastcv/src/fastcv_hal_imgproc.cpp (hunk 1 above)
Comment (quic-xuezha) [reply 3rdparty/fastcv/src/fastcv_hal_imgproc.cpp:286]: which param do you think we should add assert? or I can return CV_HAL_ERROR_NOT_IMPLEMENTED

File: 3rdparty/fastcv/src/fastcv_hal_imgproc.cpp (hunk 2 above)
Comment (quic-xuezha) [reply 3rdparty/fastcv/src/fastcv_hal_imgproc.cpp:116]: which param do you think we should add assert? or I can raise some error messages as it already in parallel we can't return CV_HAL_ERROR_NOT_IMPLEMENTED

Comment (asmorkalov) [reply 3rdparty/fastcv/src/fastcv_hal_imgproc.cpp:116]: You can use just `CV_Error()` statement then. The branch should not be reachable in normal configuration.

**Comment by asmorkalov** [comment]
@adsha-quic I merged the first part. Please rebase and fix conflicts.

**Comment by adsha-quic** [comment]
> @adsha-quic I merged the first part. Please rebase and fix conflicts.

Hey
I have rebased the change, also enabled fastcv by default

**Comment by asmorkalov** [comment]
I still see "This branch has conflicts that must be resolved" in PR status.

**Comment by adsha-quic** [comment]
> I still see "This branch has conflicts that must be resolved" in PR status.

Hey
rebased on 4.x

File: 3rdparty/fastcv/include/fastcv_hal_imgproc.hpp [hunk 3]
```diff
@@ -88,33 +93,6 @@ int fastcv_hal_sobel(
...
-    uchar*          dst_data,
-    size_t          dst_step,
-    int             width,
-    int             height,
-    int             cn,
-    double          lowThreshold,
-    double          highThreshold,
-    int             ksize,
-    bool            L2gradient);
-
```
Comment (asmorkalov) [inline 3rdparty/fastcv/include/fastcv_hal_imgproc.hpp:117]: Why is it removed? M.b. rebase issue?

File: 3rdparty/fastcv/src/fastcv_hal_imgproc.cpp [hunk 4]
```diff
@@ -743,4 +743,253 @@ int fastcv_hal_warpPerspective(
...
+
+        int dstHeight_, dstWidth_, origDstHeight_, origDstWidth_;
+        dstHeight_ = (height_ + 2*n + 1)/2;
+        dstWidth_ = (width_ + 2*n + 1)/2;
+        origDstHeight_ = (height_ + 1)/2;
+        origDstWidth_ = (width_ + 1)/2;
+
+        cv::Mat dst_padded = cv::Mat(dstHeight_, dstWidth_, CV_8U);
+
+        fcvPyramidLevel_v2 *framePyr = new fcvPyramidLevel_v2[2];
```
Comment (asmorkalov) [inline 3rdparty/fastcv/src/fastcv_hal_imgproc.cpp:786]: I propose to use local variable on stack. new is redundant here. Also it's not deleted afterwards.

File: 3rdparty/fastcv/src/fastcv_hal_imgproc.cpp [hunk 5]
```diff
@@ -743,4 +743,253 @@ int fastcv_hal_warpPerspective(
...
+                bdr = FASTCV_BORDER_REFLECT_V2;
+                break;
+            default:
+                CV_HAL_RETURN_NOT_IMPLEMENTED("border type not supported");
+        }
+
+        fcvPyramidLevel_v2 *frame1Pyr = new fcvPyramidLevel_v2[2];
+        frame1Pyr[0].ptr = NULL;
+        frame1Pyr[1].ptr = dst_data;
+        frame1Pyr[1].stride = dst_step;
```
Comment (asmorkalov) [inline 3rdparty/fastcv/src/fastcv_hal_imgproc.cpp:866]: I propose to use local variable on stack. new is redundant here. Also it's not deleted afterwards.

File: 3rdparty/fastcv/src/fastcv_hal_imgproc.cpp [hunk 6]
```diff
@@ -743,4 +743,253 @@ int fastcv_hal_warpPerspective(
...
+    size_t          dst_step,
+    int             width,
+    int             height,
+    int             depth,
+    int             scn,
+    bool            swapBlue,
+    bool            isFullRange,
+    bool            isHSV)
+{
+    if(width * height > 640 * 480)
```
Comment (asmorkalov) [inline 3rdparty/fastcv/src/fastcv_hal_imgproc.cpp:907]: Are you sure in sign? Does it work for small sizes only?

File: 3rdparty/fastcv/src/fastcv_hal_imgproc.cpp [hunk 7]
```diff
@@ -743,4 +743,253 @@ int fastcv_hal_warpPerspective(
...
+    {
+        CV_HAL_RETURN_NOT_IMPLEMENTED("input size not supported");
+    }
+    if(scn != 3 || depth != CV_8U)
+    {
+        CV_HAL_RETURN_NOT_IMPLEMENTED("src type not supported");
+    }
+    else if(isHSV != true || isFullRange != true)
```
Comment (asmorkalov) [inline 3rdparty/fastcv/src/fastcv_hal_imgproc.cpp:915]: `!isHSV || !isFullRange`?

File: 3rdparty/fastcv/src/fastcv_hal_imgproc.cpp [hunk 8]
```diff
@@ -743,4 +743,253 @@ int fastcv_hal_warpPerspective(
...
+    {
+        CV_HAL_RETURN_NOT_IMPLEMENTED("Full range HSV supported");
+    }
+    else if(swapBlue != true)
```
Comment (asmorkalov) [inline 3rdparty/fastcv/src/fastcv_hal_imgproc.cpp:919]: `!swapBlue`

File: 3rdparty/fastcv/src/fastcv_hal_imgproc.cpp [hunk 9]
```diff
@@ -743,4 +743,253 @@ int fastcv_hal_warpPerspective(
...
+{
+    if(scn != 3 || depth != CV_8U)
+    {
+        CV_HAL_RETURN_NOT_IMPLEMENTED("src type not supported");
+    }
+    else if(isCbCr != true)
+    {
+        CV_HAL_RETURN_NOT_IMPLEMENTED("CbCr supported");
+    }
+    else if(swapBlue != true)
```
Comment (asmorkalov) [inline 3rdparty/fastcv/src/fastcv_hal_imgproc.cpp:968]: The same for bools.

Comment (adsha-quic) [reply 3rdparty/fastcv/src/fastcv_hal_imgproc.cpp]: > The same for bools.

Added the suggested fixes

File: 3rdparty/fastcv/include/fastcv_hal_imgproc.hpp (hunk 3 above)
Comment (adsha-quic) [reply 3rdparty/fastcv/include/fastcv_hal_imgproc.hpp:117]: > Why is it removed? M.b. rebase issue?

It was added here by mistake, will be added in 3rd PR [opencv/opencv#26621](https://github.com/opencv/opencv/pull/26621)

**Review by asmorkalov — APPROVED** [review]
:+1: Thanks a lot for the patch! Please revert the changes in root cmake file and I'll merge the patch.

**Comment by adsha-quic** [comment]
> 👍 Thanks a lot for the patch! Please revert the changes in root cmake file and I'll merge the patch.

Hey Alex, 
done.

**Pull request merged**